### ISSUE TITLE
global: support mutable URLs in metadata

### ISSUE THREAD

**Issue by nharraud** [inveniosoftware/invenio-records#117]
This is a question I have been asking myself for some time and I wonder if we could find a common solution to this problem.

This is not a feature needed immediately. This issue's goal is to start the discussion and see if there is any easy solution. It might be even that this is not an issue at all.

**Problem Description:**
- We currently store some URLs in the record. Example: the '$schema'.
- We want to give access to every past version of a record.
- Many services want to archive records "for ever".

**Question:**
Do we want the URLs to be valid for every version of a record? If yes, do we just want to change what the REST API returns or do we also want to be able to have up to date URLs internally?

**Concrete use cases:**
- The organization running the service changes its name and its URL is changed too.
- An organization shuts down and the records are moved to another one for preservation. (Example: from one B2Share instance to another)

Do we want to still give access to the old record schemas referenced by `$schema`, and if yes how do we do it?

The situation can be quite messy if we consider that records can be reverted to old versions and have possibly invalid `$schema` URLs.

Of course there are some URLs which we can't control as they are external.

**Note:**
This can also be true for URLs inside the record's schema, as they can reference other schemas and we can store a schema inside the record itself.

**Possible solution 1:**
There could be a module which provide a *mapping* from every old url to new ones.

Pros:
- it works for external URLs too.
- it does not modify the history of a record directly. It could be called by the REST API for each request, with some caching. The replacement could be optional in order to still have access to the old version.

**Possible solution 2:**
We could support variables in the URLs, like a *HOSTNAME* variable. They would be replaced for every request.

Pros:
- simple to implement. The variables can even be replaced by the serializers. We just need to call the substitution function before retrieving fields like `$schema`.

Cons:
- this is more complicated for complex dynamic URLs and external URLs as we can't list them all as variables.

**Possible solution 3:**
If we don't care about the past version's URL internally we can just preprocess the records before sending them to the user. Those preprocessors could be called in the default invenio-records-rest serializer.

@jirikuncar @lnielsen @tiborsimko What do you think?

**Comment by lnielsen** [comment]
It's a bit hard to follow what exactly is the problem. Could you possibly try to amend with an example? I understand you want to update the URLs, but I don't understand why? 

My unfiltered thoughts:
- *Don't change history:* Past records revisions (i.e. ``record.revisions``) should not be editable in anyway. Editing the revision history would be like rewriting your entire git commit log, because an URL changed. Tampering with the history of a record would be bad record management IMHO.
- *Published JSON schemas should have very persistent URLs* which are independent of organisation renaming etc (take DataCite or Dublin Core XML schemas as examples). This could e.g. be achieved by referencing schemas using persistent identifiers such as PURLs, DOIs or EPICs.
- *Historic vs active data:* Record revisions is in my opinion history data. When you talk about changing URLs in past revisions, it makes me think that you consider the revisions as "active data".

**Comment by jirikuncar** [comment]
IMHO it's a problem of data migration. The schema can be archived in SIP/DIP store together with metadata. It's exactly the same problem as any other references to authority records.

**Comment by nharraud** [comment]
@lnielsen Yes you are right, I was thinking of revisions as 'active data'. After thinking some more it does not really make any sense as the record has been modified for a good reason and past revisions should be expected to be invalid.

I wasn't thinking of changing the history directly. Just update the metadata when a revert is done and maybe enrich on the fly the results returned by the REST API to keep it valid. But as said before, it is unnecessary.

I also agree with you when you say *Published JSON schemas should have very persistent URLs*. Without this any revert becomes dangerous. Maybe we could add a warning to invenio-jsonschema which recommends to not write its URLs in the records but use a DOI/EPIC/...
I will make sure that B2Share schemas are properly published with EPIC links.

@jirikuncar Yes and if there is any work to do during a revert it can be done on `after_record_revert` signal.

**Comment by lnielsen** [comment]
> Maybe we could add a warning to invenio-jsonschema which recommends to not write its URLs in the records but use a DOI/EPIC/...

Yes, and I would highlight it even more prominently.
